Tasks/DeployAzureResourceGroup/MachineGroupAgentExtensionManager.ts

```diff
@@ -0,0 +1,294 @@
+        }
+    }
+
+    private _settlePromise(operationParameters) {
+        if (operationParameters.failureCount + operationParameters.successCount == operationParameters.vmCount) {
+            if (operationParameters.failureCount > 0) {
+                this.log(tl.loc("MGAgentOperationOnAllVMsFailed", operationParameters.operation, ""));
+                operationParameters.deferred.reject(operationParameters.operation);
+            }
+            else {
+                if (operationParameters.vmCount !== 0) {
+                    this.log(tl.loc("MGAgentOperationOnAllVMsSucceeded", operationParameters.operation));
+                }
+                operationParameters.deferred.resolve(operationParameters.operation);
+            }
+        }
+    }
+
+    private _getExtensionDeletionCallback(operationParameters, resourceGroupName, extensionName, vmName, callback) {
+        var extensionDeletionCallback = (error, result, request, response) => {
+            if (error) {
+                this.log(tl.loc("SkipDeleteExtension", extensionName, vmName));
+                operationParameters.vmCount--;
+            }
+            else {
+                this.log(tl.loc("DeleteExtension", extensionName, vmName));
+                this._computeClient.virtualMachineExtensions.deleteMethod(resourceGroupName, vmName, extensionName, callback);
+            }
+            if (operationParameters.vmCount === 0) {
+                this._settlePromise(operationParameters);
+            }
+        }
+        return extensionDeletionCallback;
+    }
+
+    private _getPostOperationCallBack(operationParameters, extensionName, vmName) {
+        var postOperationCallBack = (error, result, request, response) => {
+            if (error) {
+                operationParameters.failureCount++;
+                operationParameters.errors += error.message;
+                operationParameters.errors += "\n";
+                this.log(tl.loc("OperationFailed", operationParameters.operation, extensionName, vmName));
+            } else {
+                operationParameters.successCount++;
+                this.log(tl.loc("OperationSucceeded", operationParameters.operation, extensionName, vmName));
+            }
+            this._settlePromise(operationParameters);
+        }
+        return postOperationCallBack;
+    }
+
+    private _getCreateOrUpdateCallback(operationParameters, extensionName, vmName, parameters, callback) {
+        var ceateOrUpdateCallBack = (error, result, request, response) => {
+            if (result) {
+                var statuses = result["properties"]["instanceView"]["statuses"];
+                for (var i = 0; i < statuses.length; i++) {
+                    var status = statuses[i]["code"].split("/");
+                    if (status.length > 1 && status[0] == "PowerState") {
+                        if (status[1] == "running") {
```

Comment:
create separate re-usable functions which take care of installing extension on running or stopped VM